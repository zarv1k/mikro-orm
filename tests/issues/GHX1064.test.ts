import { Entity, MikroORM, OneToOne, PrimaryKey, PrimaryKeyType, Property } from '@mikro-orm/core';
import { AbstractSqlDriver } from '@mikro-orm/knex';

@Entity()
export class App {

  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

}

@Entity()
export class User {

  [PrimaryKeyType]: string;
  @OneToOne({ entity: () => App, primary: true, owner: true })
  app!: App;

  @Property()
  name!: string;

}

describe('GH issue 1064', () => {

  let orm: MikroORM<AbstractSqlDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [User, App],
      dbName: 'mikro_orm_test_1064',
      type: 'postgresql',
    });
    await orm.getSchemaGenerator().ensureDatabase();
  });


  beforeEach(async () => {

    await orm.getSchemaGenerator().dropSchema();
    orm.em.clear();
  });

  afterAll(() => orm.close(true));

  it('should not create unique index on primary key column', async () => {
    const sql = await orm.getSchemaGenerator().getCreateSchemaSQL();
    expect(sql).toMatchSnapshot();
  });

  it('should not generate create unique index sql', async () => {
    await orm.getSchemaGenerator().createSchema();

    await orm.em.execute('ALTER TABLE "user"\n' +
      'DROP CONSTRAINT "user_app_id_unique"');

    const sql = await orm.getSchemaGenerator().getUpdateSchemaSQL();
    expect(sql).toMatchSnapshot();
  });

});
