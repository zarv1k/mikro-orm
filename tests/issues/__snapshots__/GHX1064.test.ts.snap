// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GH issue 1064 should not create unique index on primary key column 1`] = `
"set names 'utf8';
set session_replication_role = 'replica';

create table \\"app\\" (\\"id\\" serial primary key, \\"name\\" varchar(255) not null);

create table \\"user\\" (\\"app_id\\" int4 not null, \\"name\\" varchar(255) not null);
alter table \\"user\\" add constraint \\"user_pkey\\" primary key (\\"app_id\\");

alter table \\"user\\" add constraint \\"user_app_id_foreign\\" foreign key (\\"app_id\\") references \\"app\\" (\\"id\\") on update cascade on delete cascade;

set session_replication_role = 'origin';
"
`;

exports[`GH issue 1064 should not generate create unique index sql 1`] = `
"set names 'utf8';
set session_replication_role = 'replica';

set session_replication_role = 'origin';
"
`;
